dockerfiles/Dockerfile-rosetta

```diff
@@ -1,50 +1,309 @@
+  --depth 1 \
+  /home/opam/opam-repository \
+  && opam init --disable-sandboxing -k git -a ~/opam-repository --bare \
+  && opam switch create "${OCAML_VERSION}" "ocaml-base-compiler.${OCAML_VERSION}${OCAML_REVISION}" \
+  && opam switch "${OCAML_VERSION}"
+
+# Alternate variants for 4.07 that are included in the official opam image
+#  opam switch create 4.07+afl ocaml-variants.4.07.1+afl && \
+#  opam switch create 4.07+flambda ocaml-variants.4.07.1+flambda && \
+#  opam switch create 4.07+default-unsafe-string ocaml-variants.4.07.1+default-unsafe-string && \
+#  opam switch create 4.07+force-safe-string ocaml-variants.4.07.1+force-safe-string && \
+
+
+#################################################################################################
+# The "opam-deps" Stage
+# - Continues from the build-deps image
+# - Installs all opam dependencies and pins from coda's github
+# - Includes the entire coda codebase and submodules in "${CODA_DIR}" (must be writable by opam user)
+# - Largely mirrors/replaces ./scripts/setup-opam.sh
+#################################################################################################
+FROM build-deps AS opam-deps
+
+# location of repo used for pins and external package commits
+ARG CODA_DIR=coda
+# branch to checkout on first clone (this will be the only availible branch in the container)
+# can also be a tagged release
+ARG CODA_BRANCH=develop
+
+# location of external packages
+ARG EXTERNAL_PKG_DIR=$CODA_DIR/src/external
+
+# don't keep sources, to force reinstall of pinned packages from Coda sources
+# and to keep Docker image reasonable size
+ENV OPAMKEEPBUILDDIR false
+ENV OPAMREUSEBUILDDIR false
+
+ENV PATH "$PATH:/usr/lib/go/bin:$HOME/.cargo/bin"
+
+# git will clone into an empty dir, but this also helps us set the workdir in advance
+RUN git clone \
+  -b "${CODA_BRANCH}" \
+  --depth 1 \
+  --shallow-submodules \
+  --recurse-submodules \
+  https://github.com/CodaProtocol/coda ${HOME}/${CODA_DIR}
+
+WORKDIR $HOME/$CODA_DIR
+
+ENV OPAMYES 1
+# TODO: handle this opam work without cloning the full repository (directly pull src/opam.export)
+# Installs uri.2.2.1 and its dependencies so that build succeeds, then installs the rest of the deps
+RUN opam install \
+  uri.2.2.1 \
+  base.v0.12.0 \
+  stdio.v0.12.0 \
+  ppx_sexp_conv.v0.12.0 \
+  ppxlib.0.8.1 \
+  ocaml-compiler-libs.v0.11.0 \
+  ocaml-migrate-parsetree.1.5.0 \
+  dune.2.5.1 \
+  ocamlfind.1.8.0 \
+  base-bytes.base \
+  ppx_derivers.1.2.1 \
+  re.1.9.0 \
+  result.1.3 \
+  stringext.1.5.0 \
+  && opam switch import src/opam.export \
+  && sudo chmod -R u+rw ~/.opam
+
+RUN eval $(opam config env) \
+  && opam pin add src/external/ocaml-extlib \
+  && opam pin add src/external/ocaml-sodium \
+  && opam pin add src/external/rpc_parallel \
+  && opam pin add src/external/digestif \
+  && opam pin add src/external/async_kernel \
+  && opam pin add src/external/coda_base58 \
+  && opam pin add src/external/graphql_ppx \
```

Comment:
I don't think we should do it in this PR (maybe just make an issue) -- but it may make sense to put all the opam pins in one script and use that everywhere. I'm worried that as we keep adding these opam pins in more places we'll forget to update them everywhere whenever they change. (I think they're in a few dockerfiles, and at least one of the scripts under `scripts/`)

Reply:
I'm trying to stay away from scripts as much as possible for this as any script also requires the coda repo downloaded to use, though the submodules as they are require this anyway so I could use one. scripts/setup-opam.sh is the closest thing but that feels like overkill for this dockerfile, using a small script in both places for just pins is easy enough though.

Reply:
Noted as a todo, I think this should be grouped in with fetching and loading the src/opam.export and probably the submodules for each pin. This way we can handle all changes to the opam deps from a central place that handles fetching proper commits of each dependency, and we would have one central place to determine when to rebuild the opam-deps stage.